doc/lua_style_guide.md

```diff
@@ -0,0 +1,402 @@
+|-------------------------------------------------------------|------------------------------------------------------------------------|
+| Public API<br/>(module, function, class, method, constant)  | [LDoc][LDOC]                                                           |
+| Complex function                                            | Extract sub-functions and give them speaking names                     |
+| Complex module                                              | Extract sub-modules and give them speaking names                       |
+| Performance critical algorithm                              | This is a case where an explanation in form of a comment is acceptable |
+
+As you can see from the examples above, there is almost always a better way of documenting your code than commenting it.
+
+The only real exception are algorithms that are so performance-critical that you can't afford splitting them to make them more readable. Only after you checked that this is really necessary, you can add an explanation for the algorithm. Don't forget to justify why splitting it is not an option.
+
+While there are similar flavors to [LDoc][LDOC] (like EmmyDoc) we stick to the original in order to get the broadest tool support.
+
+Check the [LDoc handbook][LDOC] for more information on how to document the different artifacts in your code (functions, modules, classes, fields).
+
+## Scoping
+
+As in any programming language, it is best to keep the scope of all variables as narrow as possible. This prevents accidental corruption in outer scopes and makes [garbage collection](http://www.lua.org/manual/5.4/manual.html#2.5) more efficient.
+
+### Avoid the Global Scope
+
+Avoid the global scope. In Lua the global scope is the top-level scope. You either write to it by leaving out the `local` modifier and module name or by explicitly addressing the reserved `_G` variable. Globals are particular prone to name-collision and overwriting problems.
+
+That being said, there are situations where modifying the global scope actually makes sense. A very typical example is getting forward compatibility in code that has been written for Lua 5.1, but is supposed to run in Lua 5.2 or later too. The following trick maps `table.unpack` to the global scope as was the case prior to Lua 5.2:
+
+```lua
+_G.unpack = table.unpack or _G.unpack
+
+-- Old 5.1 code you depend on:
+local foo, bar = unpack(the_table)
+```
+
+A better option is available when you are in control of the code and don't depend on an external module. In this case you can turn things around to get backward compatibility instead of forward compatibility.
+
+```lua
+table.unpack = table.unpack or _G.unpack
+
+-- Old 5.1 code you depend on:
+local foo, bar = table.unpack(the_table)
+```
+
+In this variant you did not modify the global scope. Instead, you mapped a globally scoped function to the module `table`, which is a lot cleaner.
+
+### Excursion: Extending Existing Lua Modules
```

Comment:
In case you didn't know yet, this is also called [monkey patching](https://en.wikipedia.org/wiki/Monkey_patch) ;)

very rarely a good idea ;)